File: src/main/java/net/dontdrinkandroot/gitki/wicket/component/DirectoryEntriesPanel.java
package net.dontdrinkandroot.gitki.wicket.component;

import net.dontdrinkandroot.gitki.model.AbstractPath;
import net.dontdrinkandroot.gitki.model.DirectoryPath;
import net.dontdrinkandroot.gitki.model.FilePath;
import net.dontdrinkandroot.gitki.wicket.component.listitem.DirectoryListItem;
import net.dontdrinkandroot.gitki.wicket.component.listitem.FileListItem;
import net.dontdrinkandroot.wicket.bootstrap.component.list.ListGroup;
import net.dontdrinkandroot.wicket.bootstrap.component.panel.PlainPanel;
import org.apache.wicket.Component;
import org.apache.wicket.model.IModel;
import org.apache.wicket.model.Model;

import java.util.List;

/**
 * @author Philip Washington Sorst <philip@sorst.net>
 */
public class DirectoryEntriesPanel extends PlainPanel<List<AbstractPath>>
{
    public DirectoryEntriesPanel(String id, IModel<List<AbstractPath>> model)
    {
        super(id, model);
    }

    @Override
    protected Component createAfterBody(String id)
    {
        return new ListGroup<AbstractPath>(id, this.getModel())
        {
            @Override
            protected Component createListComponent(String id, IModel<AbstractPath> model)
            {
                AbstractPath path = model.getObject();
                if (path.isDirectoryPath()) {
                    IModel<DirectoryPath> pathModel = Model.of((DirectoryPath) path);
                    return new DirectoryListItem(id, pathModel);
                } else {
                    IModel<FilePath> pathModel = Model.of((FilePath) path);
                    return new FileListItem(id, pathModel);
                }
            }
        };
    }
}

File: src/main/kotlin/net/dontdrinkandroot/gitki/wicket/component/DirectoryEntriesPanel.kt
package net.dontdrinkandroot.gitki.wicket.component

import net.dontdrinkandroot.gitki.model.DirectoryPath
import net.dontdrinkandroot.gitki.model.FilePath
import net.dontdrinkandroot.gitki.model.GitkiPath
import net.dontdrinkandroot.gitki.wicket.component.listitem.DirectoryListItem
import net.dontdrinkandroot.gitki.wicket.component.listitem.FileListItem
import net.dontdrinkandroot.wicket.bootstrap.component.card.PlainCard
import net.dontdrinkandroot.wicket.bootstrap.component.list.listGroup
import org.apache.wicket.Component
import org.apache.wicket.model.IModel
import org.apache.wicket.model.Model

class DirectoryEntriesPanel(id: String, model: IModel<List<GitkiPath>>) : PlainCard<List<GitkiPath>>(id, model) {

    override fun createAfterBody(id: String): Component = listGroup(id, this.model) { id, model ->
        val path = model.getObject()
        when {
            path.directoryPath -> {
                val pathModel: IModel<DirectoryPath> = Model(path as DirectoryPath)
                DirectoryListItem(id, pathModel)
            }
            else -> {
                val pathModel: IModel<FilePath> = Model(path as FilePath)
                FileListItem(id, pathModel)
            }
        }
    }
}